sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBCheckpointFailureInjectionSuite.scala

```diff
@@ -0,0 +1,624 @@
+          var ex = intercept[SparkException] {
+            db.load(2)
+          }
+          checkError(
+            ex,
+            condition = "CANNOT_LOAD_STATE_STORE.CANNOT_READ_STREAMING_STATE_FILE",
+            parameters = Map(
+              "fileToRead" -> s"$remoteDir/2.changelog"
+            )
+          )
+
+          db.load(0)
+          FailureInjectionFileSystem.shouldFailExist = true
+          var ex2 = intercept[IOException] {
+            db.load(1)
+          }
+        }
+      }
+    }
+  }
+
+
+  /**
+   * This test is to simulate the case where a previous task had connectivity problem that couldn't
+   * be killed or write zip file. Only after the later one is successfully committed, it come back
+   * and write the zip file.
+   * The final validation isn't necessary for V1 but we just would like to make sure
+   * FailureInjectionCheckpointFileManager has correct behavior --  allows zip files to be delayed
+   * to be written, so that the test for V1 is valid.
+   */
+  test("Zip File Overwritten by Previous Task Checkpoint V1") {
+    val fmClass = "org.apache.spark.sql.execution.streaming.state." +
+      "FailureInjectionCheckpointFileManager"
+    val hadoopConf = new Configuration()
+    hadoopConf.set(STREAMING_CHECKPOINT_FILE_MANAGER_CLASS.parent.key, fmClass)
+    withTempDir { remoteDir =>
+      withSQLConf(
+        RocksDBConf.ROCKSDB_SQL_CONF_NAME_PREFIX + ".changelogCheckpointing.enabled" -> "false") {
+        val conf = RocksDBConf(StateStoreConf(SQLConf.get))
+        withDB(
+          remoteDir.getAbsolutePath,
+          version = 0,
+          conf = conf,
+          hadoopConf = hadoopConf) { db =>
+          db.put("version", "1.1")
+          commitAndGetCheckpointId(db)
+
+          db.load(1)
+          FailureInjectionFileSystem.createAtomicDelayCloseRegex = Seq(".*zip")
+          db.put("version", "2.1")
+          intercept[IOException] {
+            commitAndGetCheckpointId(db)
+          }
+
+          FailureInjectionFileSystem.createAtomicDelayCloseRegex = Seq.empty
+
+          db.load(1)
+
+          db.put("version", "2.2")
+          commitAndGetCheckpointId(db)
+
+          assert(FailureInjectionFileSystem.delayedStreams.size == 1)
+          FailureInjectionFileSystem.delayedStreams.foreach(_.close())
+        }
+        withDB(
+          remoteDir.getAbsolutePath,
+          version = 2,
+          conf = conf,
+          hadoopConf = hadoopConf) { db =>
+          // Assuming previous 2.zip overwrites, we should see the previous value.
+          assert(new String(db.get("version"), "UTF-8") == "2.1")
+        }
+      }
+    }
+   }
+
+  /**
+   * This test is to simulate the case where a previous task had connectivity problem that couldn't
+   *  be killed or write zip file. Only after the later one is successfully committed, it comes back
+   *  and write the zip file.
+   *  */
+  test("Zip File Overwritten by Previous Task Checkpoint V2") {
```

Comment:
Is this largely the same as prev test - only for checkpoint v2 ? should we parameterize into 1 test ?